### PR TITLE
Use the schema-based code generator for Go.

### PR DESCRIPTION
**Pull request by pgavlin**
This change adds a converter from an annotated Terraform schema to a
Pulumi schema, and replaces the Go code generator with a wrapper around
the Pulumi-schema-based code generator.

### CODE REVIEW

**Comment by lukehoban** [comment]
I might not be using it correctly - but when I try to use the the new code gen for the AWS provider I get:

```
go: extracting github.com/pulumi-bot/pulumi-aws v0.14.4-dev.0.20200121230752-0433c001f9c4
-> unzip /Users/luke/go/pkg/mod/cache/download/github.com/pulumi-bot/pulumi-aws/@v/v0.14.4-dev.0.20200121230752-0433c001f9c4.zip: case-insensitive file name collision: "sdk/go/aws/ec2/NetworkInterfaceAttachment" and "sdk/go/aws/ec2/networkInterfaceAttachment"
```

The issue appears to be that this code gen leads to these paths - which sounds like it causes problems for Go and/or Go modules?

```
sdk/go/aws/ec2/NetworkInterfaceAttachment/pulumiTypes.go 
sdk/go/aws/ec2/networkInterfaceAttachment/networkInterfaceAttachment.go 
```

This is with this `go.mod`:

```
module lukehoban/go-simple

go 1.13

require (
	github.com/pulumi/pulumi v1.8.2-0.20200121224548-d400f2683278
	github.com/pulumi/pulumi-aws v1.18.1-0.20200116192439-d1b5c8a4765a
)

replace github.com/pulumi/pulumi-aws => github.com/pulumi-bot/pulumi-aws v0.14.4-dev.0.20200121230752-0433c001f9c4
```

(no idea why this resolved to `v0.14.4-*` - but the commit hash is right which I believe is all that matters - and the error is definitely related to the code generated here.

**Comment by pgavlin** [comment]
> and the error is definitely related to the code generated here.

Yeah, the code generator is generating filenames with casing that is inconsistent with the old code generator. We can keep the original casing or we can delete the existing SDKs and regenerate them (i.e. `git rm -rf sdk/go/* && git commit -a -m "remove old Go SDK" && make default`).

**Comment by pgavlin** [comment]
Hm, something's a bit off here in general: I'm seeing unexpected subpackages in the AWS output . I'll see what changes need to be made in the code generator to fix.

**Comment by lukehoban** [comment]
Hmm - but at least the diff in https://github.com/pulumi-bot/pulumi-aws/compare/fb8c334f0..0433c001f shows both these paths as being new files.  Is that wrong due to some aspect of how this merge is happening?  

Either way - if it's a problem we will be able to fix when we bring these changes into the providers for real then great.

**Comment by pgavlin** [comment]
I grabbed the wrong copy of `generate_schema.go`. 🤦‍♂

File: pkg/tfgen/generate_schema.go [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+
+	// Override the nested type name, if necessary.
+	if typ.nestedType.Name().String() != "" {
+		baseName = typ.nestedType.Name().String()
+	}
+
+	typeName := baseName + strings.Title(nameSuffix)
+	typ.name = typeName
+
+	//contract.Assertf(nt.nameToType[typeName] == nil, "nested type %q already declared", typeName)
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply pkg/tfgen/generate_schema.go]: Yep, good catch.

File: pkg/tfgen/generate_schema.go [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+		description = g.genRawDocComment(prop.rawdoc)
+	}
+
+	language := map[string]json.RawMessage{}
+	if prop.info != nil && prop.info.CSharpName != "" {
+		language["csharp"] = rawMessage(map[string]string{"name": prop.info.CSharpName})
+	}
+
+	// Due to bugs in earlier versions of the bridge, we want to keep the Python code generator from case-mapping
+	// properties an object-typed element that are not Map types. This is consistent with the earlier behavior.
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply pkg/tfgen/generate_schema.go]: Yes.

**Comment by pgavlin** [comment]
> It looks like we're relying on end to end code generation testing to validate the schema generation code. Do you think this is sufficient? Would some unit test coverage bring greater confidence that this is working as intended?

This is correct. It would be great to get some unit test coverage, but doing so has historically been quite difficult or tedious, as it requires authoring non-trivial TF schemas. I'm not going to add that as part of this PR, but we should consider doing this in the future. This should be a bit simpler now, as we really just need to test the TF schema -> Pulumi schema conversion (the codegen unit tests can live in pulumi/pulumi).

**Comment by lukehoban** [comment]
I'm seeing a different kind of name collision now - this one perhaps is just due to needing to do `git rm`?

```
case-insensitive file name collision: "sdk/go/aws/datasync/s3Location.go" and "sdk/go/aws/datasync/s3location.go"
```

**Comment by pgavlin** [comment]
> I'm seeing a different kind of name collision now - this one perhaps is just due to needing to do git rm?

I would try that, yes.

What platform are you on?

**Comment by lukehoban** [comment]
> doing so has historically been quite difficult or tedious, as it requires authoring non-trivial TF schemas. I'm not going to add that as part of this PR, but we should consider doing this in the future. This should be a bit simpler now, as we really just need to test the TF schema -> Pulumi schema conversion (the codegen unit tests can live in pulumi/pulumi).

FWIW - this is tracked in https://github.com/pulumi/pulumi-terraform-bridge/issues/34 - which has been open for 2.5 years!

**Comment by lukehoban** [comment]
> I would try that, yes.

I'm currently relying on the `pulumi-bot` fork it posts above.  I guess I'd have to manually build a branch with these changes - I'll try that.

> What platform are you on?

macOS

File: pkg/tfgen/generate_schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukehoban) [reply pkg/tfgen/generate_schema.go]: Do we have an issue tracking this specifically - if not we should open one and mark it as `2.0` and `Breaking Change`.

**Comment by pgavlin** [comment]
> > What platform are you on?
> 
> macOS

Got it. I haven't seen those issues when building locally, but I also haven't tried using the pulumi-bot fork.

File: pkg/tfgen/generate_schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pgavlin) [reply pkg/tfgen/generate_schema.go]: https://github.com/pulumi/pulumi/issues/3151 is the closest we have.

[the remark this answers is not in the document]
Comment (pgavlin) [reply pkg/tfgen/generate_schema.go]: (added a link to the bug in the comment)

**Pull request merged**